Expose refmt

glennsl opened the pull request.

bassjacob commented on bin/refmt:
```diff
@@ -0,0 +1,14 @@
+
+try {
+    child_process.execFileSync(exe, delegate_args, { stdio: 'inherit' })
+} catch (e) {
```
is there any value in surfacing what the error was, even just as far as the message?

bobzhang commented:
cc @chenglou @jordwalke -- I think the consensus was not exposing refmt?

jordwalke commented:
The one risk is that people begin to rely upon it and then it can never be removed. Longer term it will make more sense to have a separate prebuilt binary package just for refmt, so that you can have one version of BuckleScript that can work with any version of refmt. Here's the scenario that solves:

1. Bucklescript makes a large improvement, or bug fix.
2. You have two projects, one that depends on reason 1.13, and another that depends on 1.14.
3. You want both projects to have the BuckleScript improvement, but you cannot upgrade the 1.13 project due to lack of time, or because it's a UI component written by someone else.

Can I learn more about the use case this solves?

glennsl commented:
I'm using it for a tool that extracts and compiles code blocks in markdown files, using a command like this: https://github.com/glennsl/checkmarked/blob/master/test/checkmarked.config.json#L7

But there are two problems with that:
1. It should use the parent project's build setup if available
2. `refmt`s location depends on where bs-platform happens to be in the dependency tree (and of course the fact that it's internal provides no guarantees)

If it's not exposed I'd need a wrapper script to look it up, and still it would be prone to fail if something changes. Could we expose it under a different name, to make it clear that it's internal or "special"? E.g. `internal_refmt` or `bs-refmt`.

chenglou commented:
As an aside: great idea for a tool =)
I'd run in on our docs

bobzhang commented:
so what's the conclusion? I don't have strong opinions on this issue

jordwalke commented:
If reason-cli ends up not working well enough for this, and you do want to expose refmt from bs-platform, I'd suggest somehow marking it as "temporary" or "internal" so that bucklescript doesn't need to remain on the hook for supporting the binaries indefinitely.
I would like to eventually get some prebuilt binaries just for refmt that allow us to include an arbitrary refmt version with an arbitrary bs-platform version. I understand if you feel like using bs-platform's internal one until then, but if so I think it's important to either clearly mark it as not-entirely-supported (as you suggested with a name) or find a volunteer that is explicitly willing to support it indefinitely.

glennsl commented:
reason-cli is a bit of a heavy dependency, and wouldn't test the actual build setup anyway, so I've renamed it `internal_refmt`, which I hope should make the intention clear, and removed the changelog entry.

As for examples, it really just extracts code blocks from markdown files, write them out to separate files and runs a specified command on each of them based on the language of the given code block. Typically just compilation, but you could easily write a test script to run them and verify the output is as expected too, for example.

bobzhang commented:
what do you think? I am fine with it

jordwalke commented:
I don't feel strongly. @glennsl, was your intent that `internal_refmt` conveys that the binary is not part of the "official" contract of the bs-platform package and that it may be refactored into a standalone package?

glennsl commented:
@jordwalke Yes, that's my intent (and hope!)

bobzhang commented:
how about renaming it into bsrefmt?

glennsl commented:
@bobzhang Done. I figured `internal_refmt` was less committing, but `bsrefmt` is cool.
